Formatting

frigus02 opened the pull request:
Enable formatting support using pgFormatter (requires Perl). Fixes #4 

To do before merge:

- [x] Doesn't work in VS Code at the moment. That might be because VS Code doesn't use the TypeScript language server.
- [x] Placeholder substitutions (applied before formatting) need to be reverted before the text is replaced.
- [x] Need to apply proper indentation after formatting.
- [x] pgFormatter seems to require `HOME` to be set, which isn't the case for Windows. Can we work around this?

Known issues:

- Formatting seems to be quite slow in the index.ts file in the test project. Might be because of the amount of SQL tags. Not sure if we can speed this up. But formatting can be disabled, so I think this is fine.
- In VS Code Formatting only works using the "TypeScript and JavaScript language features" formatter. If you use another formatter (e.g. Prettier), it doesn't work.

frigus02 committed:
First attempt at format support

frigus02 committed:
Enable formatting in test project

frigus02 committed:
Make it work for VS Code

- Enable formatting by default and detect Perl
- Add note to README explaining multiple tsserver gotcha with
  tsconfig.json file installation method

frigus02 committed:
Fix err if SQL has more params than TS expressions

If a SQL statement contains a parameter with a number greater than the
number of template expressions in the template literal, the code used to
throw a "cannot read X of undefined" error.

This fixes it by ignoring those parameters.

frigus02 commented:
Waiting to see if the options proposed in https://github.com/gajus/pg-formatter/pull/7 are acceptable. Those would help with the `HOME` problem and with honoring the tabs/spaces setting.

frigus02 committed:
Honor tabs/spaces setting

frigus02 commented:
Gajus merged this and released a new version super quickly 🎉

karlhorky commented:
As for matching the indentation, I wrote a first version for `eslint-plugin-sql` here, maybe it can offer some inspiration:

https://github.com/gajus/eslint-plugin-sql/pull/11

frigus02 committed:
Extract parameter locations

frigus02 committed:
Exclude template expressions for formatting edits

Return spans of formatted SQL, that don't include the template
expressions. This ensures that expressions stay as they are and they can
even be formatted in the same go by the TypeScript formatter.

frigus02 commented:
Ah, I just realized the logic to find the first node in the line doesn't work with the TypeScript AST for code like this:

```ts
const {
    one
} = sql`
    SELECT 1
`;
```

The first character on the line is `}`, which belongs to the variable declaration node "{ one }", which starts 2 lines above.

The TypeScript language service has a function `getIndentationAtPosition`. I hope we can use that.

frigus02 committed:
Indent SQL matching to template literal

frigus02 committed:
Count line indentation manually

The language service `getIndentationAtPosition` returned different
values for the following two code snippets:

    // Indentation returned for starting `: 4
    const query = sql("delete-orders")`
        DELETE FROM orders
    `;

    // Indentation returned for starting `: 0
    const query = sql`
        DELETE FROM orders
    `;

I couldn't figure out why this is the case. This changes indentation
detection to look at the code in the line and count leading spaces/tabs.
The logic for this has been copied from TypeScript's implementation of
the `getFormattingEditsAfterKeystroke` function.

karlhorky commented:
Oh wow, manually counting in a5c8e2d0cee21bcdc2a683ac0f870f05ba5784b7 ! Sorry about my not-quite-so-working solution with the AST! 😨 I guess there are some nasty edge cases here...

frigus02 committed:
Clean up

frigus02 commented:
No worries. I borrowed the rest of your code and adjusted it to the TypeScript format options. Thanks a lot for your help on this. I probably wouldn't have done this otherwise.

karlhorky commented:
Cool, really awesome work again! Super fast execution and working through some really complex issues! 💯
